drivers/bt_ll_nrfxlib/entropy/entropy_ll_nrfxlib.c

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <entropy.h>
+#include <soc.h>
+#include <device.h>
+#include <irq.h>
+#include <kernel.h>
+#include <kernel_includes.h>
+#include <ble_controller.h>
+#include <ble_controller_soc.h>
+
+#include <multithreading_lock.h>
+
+/* The RNG driver data. */
+struct rng_driver_data {
+	/* Used to fill the pending client buffer with new random values after
+	 * RNG IRQ is triggered.
+	 */
+	struct k_sem sem_sync;
+};
+
+static struct rng_driver_data rng_data;
+
+static inline struct rng_driver_data *rng_driver_data_get(struct device *dev)
+{
+	__ASSERT_NO_MSG(dev != NULL);
+	__ASSERT_NO_MSG((intptr_t) dev->driver_data == (intptr_t) &rng_data);
+
+	return (struct rng_driver_data *) (dev)->driver_data;
+}
+
+static int rng_driver_get_entropy(struct device *dev, u8_t *buf, u16_t len)
+{
+	struct rng_driver_data *rng_dev = rng_driver_data_get(dev);
+	u8_t *p_dst = buf;
+	u32_t bytes_left = len;
+
+	while (bytes_left > 0) {
+		u32_t bytes_read = 0;
+
+		while (bytes_read == 0) {
+			int errcode = MULTITHREADING_LOCK_ACQUIRE();
+
+			if (errcode) {
+				return errcode;
+			}
+			bytes_read = ble_controller_rand_vector_get(p_dst,
+								    bytes_left);
+			MULTITHREADING_LOCK_RELEASE();
+
+			if (!bytes_read) {
+				/* Put the thread on wait until next interrupt
+				 * to get more random values.
+				 */
+				k_sem_take(&rng_dev->sem_sync, K_FOREVER);
+			}
+		}
+
+		p_dst += bytes_read;
+		bytes_left -= bytes_read;
+	}
+
+	return 0;
+}
+
+static int rng_driver_get_entropy_isr(struct device *dev, u8_t *buf, u16_t len,
+				      u32_t flags)
+{
+
+	int errcode = 0;
+
+	if (likely((flags & ENTROPY_BUSYWAIT) == 0)) {
+		errcode = MULTITHREADING_LOCK_ACQUIRE_NO_WAIT();
```

Comment:
Usage of lock shared with thread could cause starvation of the ISRs. Please also note that this function should be re-entrant and caller has possibility to handle errors.

Reply:
This is not possible at the current moment. We need to accept possibility of starvation for this release and document the change in behavior when using the BT_LL_NRFXLIB.

Reply:
Documented here: https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/687/files#diff-3408bae23cb79d87a923a686e0dd82cf